File: checker/script/analyze/analyzer/analysis_workflow_executor.py
from typing import List

from analyzer.clang_static_analyzer import run_clang
from analyzer.timer import Timer
from domain.analysis import Analysis
from domain.analysis_detail import AnalysisDetail
from domain.value.file import File


class AnalysisWorkflowExecutor():
    def __init__(self, files: List[File]) -> None:
        self.file = files
        self.analysis = Analysis(count_files=len(files))
        self.analysis_detail_list: List[AnalysisDetail] = []

    def __analyze(self) -> List[AnalysisDetail]:
        analysis_detail_list: List[AnalysisDetail] = []
        for file in self.file:
            analysis_detail_list.extend(run_clang(file))
        return analysis_detail_list

    def run(self):
        timer = Timer()
        timer.start()

        self.analysis_detail_list = self.__analyze()

        timer.stop()
        self.analysis.set_processing_time(timer.get_time())

    def get_analysis(self) -> Analysis:
        return self.analysis

    def get_analysis_detail_list(self) -> List[AnalysisDetail]:
        return self.analysis_detail_list


if __name__ == '__main__':
    analyzer = AnalysisWorkflowExecutor([File("path/to/file")])

    analyzer.run()

    print(analyzer.get_analysis())
    print(analyzer.get_analysis_detail_list())

File: checker/reporter/src/analyzer/analysis_workflow_executor.rs
use crate::{
    domain::{analysis::Analysis, analysis_detail::AnalysisDetail, value::file::File},
    y2k38_checker::y2k38_checker_trait::Y2k38Checker,
};

use super::timer::Timer;

pub struct AnalysisWorkflowExecutor {
    files: Vec<File>,
    analysis: Analysis,
    analysis_details: Vec<AnalysisDetail>,
    y2k38_checker: Box<dyn Y2k38Checker>,
}

impl AnalysisWorkflowExecutor {
    pub fn new(files: Vec<File>, y2k38_checker: Box<dyn Y2k38Checker>) -> Self {
        AnalysisWorkflowExecutor {
            files,
            analysis: Analysis::new(),
            analysis_details: Vec::new(),
            y2k38_checker,
        }
    }

    pub fn run(&mut self) {
        let mut timer = Timer::new();
        timer.start();

        self.analysis_details = self
            .files
            .iter()
            .flat_map(|file| run_checker(file, &*self.y2k38_checker))
            .collect();

        timer.stop();
        self.analysis.set_processing_time(timer.time());
    }

    pub fn files(&self) -> &Vec<File> {
        &self.files
    }
    pub fn analysis(&self) -> &Analysis {
        &self.analysis
    }
    pub fn analysis_detail(&self) -> &Vec<AnalysisDetail> {
        &self.analysis_details
    }
}

fn run_checker(file: &File, y2k38_checker: &dyn Y2k38Checker) -> Vec<AnalysisDetail> {
    let result = y2k38_checker.run(file, false);
    if let Ok(analysis_details) = result {
        analysis_details
    } else {
        println!("[Error] in Y2k38Checker.run(): {:?}", result.err());
        println!("- file: {:#?}", file.path());
        println!("- Y2k38Checker: {:#?}", y2k38_checker.name());
        println!("- description: {:#?}", y2k38_checker.description());
        Vec::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::y2k38_checker::clang_plugin_y2k38_checker::ClangPluginY2k38Checker;

    #[test]
    fn test_run() {
        let file = File::new(String::from(
            "/root/y2k38-checker/dataset/blacklist/read-fs-timestamp.c",
        ));
        let checker = ClangPluginY2k38Checker {};
        let mut executor = AnalysisWorkflowExecutor::new(vec![file], Box::new(checker));
        executor.run();

        assert!(!executor.analysis_detail().is_empty());
    }
}
